### PR TITLE
Basic building blocks.

### PR DESCRIPTION
**Pull request by skywhale**

### CODE REVIEW

**Comment by bschwind** [comment]
Like you mentioned before, it would probably be more Rust-y to use generics for the ship but trait objects aren't bad either!

I need to look a little more closely at the rest of the code, but I think I get the main idea of it, and it looks good!

File: client/src/ship.rs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+impl Engine for CustomEngine {
+    fn throttle(&mut self) -> f32 {
+        self.throttle
+    }
+}
+
+// TODO(player): Pick a right set of components to create a custom engine.
+// TODO(ryo): Implement predefined components that can be combined with iterator.
+impl CustomEngine {
+    pub fn with<F1, F2, F3, IT>(component1: F1, component2: F2, component3: F3) -> Self
```
Comment (bschwind) [inline client/src/ship.rs:150]: Could you explain this one a bit more? I remember you said something about the server having a secret which is a result of using the "right' set of components, but how will the player know which ones are right? Are we giving particular clues that will help them?

Comment (skywhale) [reply client/src/ship.rs:150]: the original idea was for me to provide several pre-defined functions, and players to choose 3 of them to combine. some combinations don't even compile, but among the ones that compile, they need to try and see what happens.

but after chatting a bit with Jake, I'm thinking to simplify it a bit. One idea is that the engine has non-linear output to the throttle input e.g. https://bit.ly/2NUwbPP

A lot of players will probably start with a constant of 1.0 for the throttle, but you can maximize the engine output by using argmax of that function. Using iterators is one way to solve it.

**Comment by skywhale** [comment]
Thank you Brian. I will see if I can use generics at least for the Ship.

**Pull request merged**